From: Olivier Fourdan <ofourdan@redhat.com>
Date: Wed, 25 May 2016 15:04:27 +0200
Subject: wayland: add extended state for tiled

xdg-shell allows desktop environments to extend the list of states
within a given range.

Use this possibility to add a new state for tiled so that gtk+ can
benefit from this.

Bug: https://bugzilla.gnome.org/show_bug.cgi?id=766860
Origin: upstream, 3.21.3, commit:e7430a45356da043019659db9b4892a08f889c7e
[smcv: backported to 3.20.x]
---
 src/wayland/meta-wayland-surface.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

Index: mutter-3.20.3/src/wayland/meta-wayland-surface.c
===================================================================
--- mutter-3.20.3.orig/src/wayland/meta-wayland-surface.c	2016-06-30 00:45:00.610896657 +0200
+++ mutter-3.20.3/src/wayland/meta-wayland-surface.c	2016-06-30 00:45:00.610896657 +0200
@@ -56,6 +56,13 @@
 #include "meta-surface-actor-wayland.h"
 #include "meta-xwayland-private.h"
 
+/*
+ * Define GNOME additional states to xdg-shell
+ * The current reserved range for GNOME is 0x1000 - 0x1FFF
+ */
+
+#define XDG_SURFACE_STATE_GNOME_TILED 0x1000
+
 enum {
   PENDING_STATE_SIGNAL_APPLIED,
 
@@ -2514,6 +2521,13 @@
       s = wl_array_add (states, sizeof *s);
       *s = XDG_SURFACE_STATE_ACTIVATED;
     }
+  /* GNOME extension to xdg-shell states */
+  if (window->tile_mode == META_TILE_LEFT ||
+      window->tile_mode == META_TILE_RIGHT)
+    {
+      s = wl_array_add (states, sizeof *s);
+      *s = XDG_SURFACE_STATE_GNOME_TILED;
+    }
 }
 
 void
